hathor/indexes/rocksdb_deps_index.py

```diff
@@ -0,0 +1,329 @@
+            # add at least one dependency, otherwise this tx should be full validated
+            if not tx_storage.transaction_exists(tx_dep_hash):
+                self.log.debug('tx parent is needed', tx=tx.hash.hex(), tx_dep=tx_dep_hash.hex())
+                batch.put((self._cf, self._to_key_needed(tx_dep_hash)), self._to_value_needed(height, tx.hash))
+
+    def remove_ready_for_validation(self, tx: bytes) -> None:
+        self._db.delete((self._cf, self._to_key_ready(tx)))
+
+    def next_ready_for_validation(self, tx_storage: 'TransactionStorage', *, dry_run: bool = False) -> Iterator[bytes]:
+        import rocksdb
+        batch = rocksdb.WriteBatch()
+        ready = self._drain_all_sorted_ready(tx_storage, batch)
+        if not dry_run:
+            self._db.write(batch)
+        while ready:
+            yield from ready
+            batch = rocksdb.WriteBatch()
+            ready = self._drain_all_sorted_ready(tx_storage, batch)
+            if not dry_run:
+                self._db.write(batch)
+
+    def _drain_all_sorted_ready(self, tx_storage: 'TransactionStorage', batch: 'rocksdb.WriteBatch') -> List[bytes]:
+        ready = list(self._drain_all_ready(tx_storage, batch))
+        ready.sort(key=lambda tx_hash: tx_storage.get_transaction(tx_hash).timestamp)
+        return ready
+
+    def _drain_all_ready(self, tx_storage: 'TransactionStorage', batch: 'rocksdb.WriteBatch') -> Iterator[bytes]:
+        it = self._db.iterkeys(self._cf)
+        seek_key = self._to_key_ready()
+        self.log.debug('seek to start', seek_key=seek_key.hex())
+        it.seek(seek_key)
+        for _, key in it:
+            key_any = self._from_key_any(key)
+            if key_any.tag is not _Tag.READY:
+                break
+            tx_hash = key_any.tx_hash
+            assert tx_hash is not None
+            self.log.debug('seek found', tx=tx_hash.hex())
+            batch.delete((self._cf, key))
+            yield tx_hash
+        self.log.debug('seek end')
+
+    def iter(self) -> Iterator[bytes]:
+        yield from self._iter_has_rev_deps()
+
+    def _iter_has_rev_deps(self) -> Iterator[bytes]:
+        it = self._db.iterkeys(self._cf)
+        seek_key = self._to_key_reverse()
+        self.log.debug('seek to start', seek_key=seek_key.hex())
+        it.seek(seek_key)
+        for _, key in it:
+            key_any = self._from_key_any(key)
+            if key_any.tag is not _Tag.REVERSE:
+                break
+            tx_dep_hash = key_any.tx_dep_hash
+            assert tx_dep_hash is not None
+            self.log.debug('seek found', tx_dep=tx_dep_hash.hex())
+            yield tx_dep_hash
+            # XXX: do this seek to skip directly to the next tx_dep_hash, otherwise we would have to iterate until the
+            #      found key has a different tx_dep_hash
+            seek_key = incr_key(self._to_key_reverse(tx_dep_hash))
```

Comment:
What does `incr_key(...)` do?

Reply:
`incr_key(x: bytes)` will basically treat `x` as an integer and increment it by 1, it mainly does byte-by-byte increments dealing with overflows, the main usefulness is when navigating keys that are like this:

Consider a key that is `[2 bytes hash][2 bytes timestamp]`, with entries:

- `00feffff`
- `00ff0001`
- `00ff0002`
- `00ffffff`
- `01000000`

If you want all transactions with hash `00ff` in _ascending_ order, you have to use a "forward" iterator, seek to `00ff` and start from there (the first yield will be `00ff0001`), if you want a all of the same transactions in _descending_ order, you have to use a "reverse" iterator, "seek for the prev" of `0100` and start from there (the first yield will be `00ffffff`). This is done a lot in `rocksdb_timestamp_index` and it can be easier to understand taking a look at that implementation.

Reply:
Got it. Thanks for the explanation. In which byte does it start the incrementing? The first or the last byte? For example, `incr_key(b'abcde')` would be `bbcde` or `abcdf`?

I feel we should add it to the comments. Not the forward and reverse iterator part, but why we are using `incr_key(...)` in the first place.

Reply:
`incr_key(b'abcde') -> b'abcdf'`, which is how rocksdb sorts keys, the docstring of `incr_key` is fairly complete IMO

the XXX comment above is the explanation, but maybe it wasn't clear enough

Reply:
Got it. Maybe you can state that you will do a reverse traversal (from end to beginning).

Reply:
But it is not a reverse traversal though, `self._to_key_reverse(tx_dep_hash)` is to format `tx_dep_hash` as a key to get its _reverse dependency_. The `incr_key` is used to skip everything under `tx_dep_hash`, which is to say, if it was some list like my previous comment, if `tx_dep_hash=b'00ff'`, the skip will make it so the iterator will get `01000000` on the next iteration (of which we extract `0100` from). That way we don't need to visit `00ff0002` and `00ffffff` and skip them manually, we can let rocksdb's iterator engine skip them.

Reply:
I finally understood how it works. Thanks for clarifying. We need to double check that `it.seek(...)` simply walks through the iterator until it finds a match. Otherwise, this method would be `O(m log n)`.

Reply:
We might just leave a note and investigate it further it we have performance issues here. I'm good to go with the current implementation.